File: packages/sdk-ts/src/kernel/lock-manager.ts
/**
 * LockManager — namespace class extracting lockfile logic
 * from ManifestInstance.
 *
 * Usage: `mi.lock.generate()` — equivalent to `mi.generateLock()`.
 *
 * This is an extraction (Chunk 2 of the kernel simplification plan).
 * The original methods on ManifestInstance are preserved; both APIs
 * return identical results.
 */

import { documentHash, type LockEntry, type Lockfile } from "./lock.js";
import type { ManifestInstance } from "./instance.js";

// ---------------------------------------------------------------------------
// LockManager
// ---------------------------------------------------------------------------

export class LockManager {
  constructor(private host: ManifestInstance) {}

  /**
   * Generate a lockfile snapshot from the current documents.
   * Equivalent to `mi.generateLock()`.
   */
  generate(): Lockfile {
    const entries: LockEntry[] = this.host.documents.map((d) => {
      const sha = documentHash(d.raw);
      return {
        name: d.name,
        kind: d.kind,
        apiVersion: d.apiVersion,
        origin: d.origin,
        path: "",
        sha256: sha,
      };
    });

    return {
      scope: this.host.scope,
      documents: entries,
      lockVersion: 3,
      generatedAt: new Date().toISOString(),
    };
  }
}

File: packages/sdk-py/dna/kernel/lock_manager.py
"""LockManager — mi.lock.generate() namespace.

Extracts lockfile logic from ManifestInstance. Both
``mi.generate_lock()`` and ``mi.lock.generate()`` return identical results.
"""
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from dna.kernel.instance import ManifestInstance
    from dna.kernel.lock import Lockfile


class LockManager:
    """Namespace for lockfile operations — accessed via ``mi.lock``."""

    def __init__(self, host: ManifestInstance) -> None:
        self._host = host

    def generate(self) -> Lockfile:
        """Generate lockfile with SHA256 per document.

        Equivalent to ``mi.generate_lock()``.
        """
        import hashlib
        import json
        from dna.kernel.lock import LockEntry, Lockfile

        entries = []
        for d in self._host.documents:
            raw_json = json.dumps(d.raw, sort_keys=True, ensure_ascii=False)
            sha = hashlib.sha256(raw_json.encode()).hexdigest()
            entries.append(LockEntry(
                name=d.name, kind=d.kind, api_version=d.api_version,
                origin=getattr(d, "origin", "local"),
                path="",
                sha256=sha,
            ))
        return Lockfile(scope=self._host.scope, documents=entries)
